salt/modules/win_event_viewer.py

```diff
@@ -0,0 +1,285 @@
+        data = tuple(new_list)
+    elif isinstance(data, str):
+        # data is turning into bytes because if was a string
+        data = data.encode(encoding)
+
+    return data
+
+
+def _get_raw_time(time):
+    '''
+    info: will make a pywintypes.datetime into a tuple
+    :param time: pywintypes.datetime
+    :return: tuple
+    '''
+
+    return time.year, time.month, time.day, time.hour, time.minute, time.second
+
+
+def make_event_dict(event):
+    '''
+    info: will make a PyEventLogRecord into a dict
+    :param event: PyEventLogRecord
+    :return: dict
+    '''
+
+    event_dict = {}
+    for event_part in EVENT_PARTS:
+        # get object value and add it to the event dict
+        event_dict[event_part] = getattr(event, event_part[0].upper() + event_part[1:], None)
+
+    # format items
+    event_dict['eventID'] = winerror.HRESULT_CODE(event_dict.get('eventID'))
+    if event_dict.get('sid') is not None:
+        event_dict['sid'] = event_dict.get('sid').GetSidIdentifierAuthority()
+    event_dict['timeGenerated'] = _get_raw_time(event_dict.get('timeGenerated'))
+    event_dict['timeWritten'] = _get_raw_time(event_dict.get('timeWritten'))
+
+    return _change_str_to_bytes(event_dict)
+
+
+def _get_event_handler(log_name, target_computer=None):
+    '''
+    info: will try to open a PyHANDLE
+    :param log_name: str
+    :param target_computer: None or str
+    :return: PyHANDLE
+    '''
+
+    try:
+        return win32evtlog.OpenEventLog(target_computer, log_name)
+    except pywintypes.error:
+        raise FileNotFoundError('Log "{0}" of "{1}" can not be found or access was denied!'.format(log_name,
+                                                                                                   target_computer))
+
+
+def _close_event_handler(handler):
+    '''
+    info: will close the event handler
+    :param handler: PyHANDLE
+    :return:
+    '''
+
+    win32evtlog.CloseEventLog(handler)
+
+
+def get_event_generator(log_name, target_computer=None, raw=False):
+    '''
+    info: will get all log events one by one
+    warning: events are not in exact order
+    :param log_name: str
+    :param target_computer: None or str
+    :param raw: bool: True: PyEventLogRecord False: dict
+    :return: PyEventLogRecord or dict
+    '''
+
+    handler = _get_event_handler(log_name, target_computer)
+    flags = win32evtlog.EVENTLOG_BACKWARDS_READ | win32evtlog.EVENTLOG_SEQUENTIAL_READ
+    event_count = 0
+
+    while win32evtlog.GetNumberOfEventLogRecords(handler) > event_count:
+        # get list of some of the events
+        events = win32evtlog.ReadEventLog(handler, flags, 0)
+        if not events:
+            # event log was updated and events are not ready to be given yet
+            # rather than wait just return
+            break
+        for event in events:
+            event_count += 1
+            if raw:
+                yield event
+            else:
+                yield make_event_dict(event)
```

Comment:
Does `events` not return all of the available events? It seems like you could simplify this function by just changing it to:

`for event in win32evtlog.ReadEventLog(handler, flags, 0):`

Or am I missing something? I do know that sometimes there are some weird Windows things :trollface:

Reply:
win32evtlog.ReadEventLog(handler, flags, 0) does not give you all of the events at once. You most keep calling it until you get an empty list. If you don't keep count of events given win32evtlog.ReadEventLog will hang until a new event is add it.

Reply:
The behavior of  win32evtlog.ReadEventLog is very funny
it give a list of events but not all of the events 
if it gives an empty list a new event was given to the log but is not ready
if it hangs its waiting for a event to be given

Reply:
Ahh. I see now. Looks good 👍